[stdlib]  Fix atol parsing for prefixed integer literals with leading underscores

jjvraw opened the pull request:
This addresses #3182 to correctly parse prefixed string literals. Changes include:
- Bug-fix
- Added appropriate test cases to verify the new behaviour

When the base 2, 8, 16 are set or inferred (base 0), for a string prefixed appropriately, leading underscores should be allowed as per Python's integer literal grammar: 

```
integer      ::=  decinteger | bininteger | octinteger | hexinteger
decinteger   ::=  nonzerodigit (["_"] digit)* | "0"+ (["_"] "0")*
bininteger   ::=  "0" ("b" | "B") (["_"] bindigit)+
octinteger   ::=  "0" ("o" | "O") (["_"] octdigit)+
hexinteger   ::=  "0" ("x" | "X") (["_"] hexdigit)+
nonzerodigit ::=  "1"..."9"
digit        ::=  "0"..."9"
bindigit     ::=  "0" | "1"
octdigit     ::=  "0"..."7"
hexdigit     ::=  digit | "a"..."f" | "A"..."F"
```

Currently, this such a string passed to `atol` raises. The changes in this PR fixes this issue. 

I should note that I have another PR (#3207) to refactor, and clean up, `atol` for reusability to introduce `stol` as per request #2639. If this PR is merged, I will update #3207, which will require a minor tweak.

martinvuyk reviewed (approved):
LGTM!
btw. loved this refactor
```mojo
    var start = pos
    if start + 1 < str_len:
        var prefix_char = str_ref[start + 1]
        if str_ref[start] == "0" and (
            (base == 2 and (prefix_char == "b" or prefix_char == "B"))
            or (base == 8 and (prefix_char == "o" or prefix_char == "O"))
            or (base == 16 and (prefix_char == "x" or prefix_char == "X"))
        ):
            start += 2
    return start
```

jjvraw committed:
atol bug-fix for leading underscores

- Support leading underscores (bug fix) for prefixed literal
- Added relevant tests

Signed-off-by: Joshua James Venter <venter.joshua@gmail.com>
